src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -560,16 +560,26 @@ void MacroAssembler::call_VM_helper(Register oop_result, address entry_point, in
 
 // Maybe emit a call via a trampoline.  If the code cache is small
 // trampolines won't be emitted.
-
 address MacroAssembler::trampoline_call(Address entry, CodeBuffer* cbuf) {
   assert(JavaThread::current()->is_Compiler_thread(), "just checking");
   assert(entry.rspec().type() == relocInfo::runtime_call_type
          || entry.rspec().type() == relocInfo::opt_virtual_call_type
          || entry.rspec().type() == relocInfo::static_call_type
          || entry.rspec().type() == relocInfo::virtual_call_type, "wrong reloc type");
+  assert(CodeCache::contains(entry.target()), "target outside CodeCache");
+
+  // Runtime calls are statically bound.
+  // Once they are generated a call site cannot be changed.
+  // Use target_needs_far_branch to check whether a far branch is needed.
+  // For Java calls a callee can be changed at any time as a result of
+  // callee deoptimization or the callee being C1 compiled has become C2 compiled.
+  // If CodeCache size > 128M, such Java calls must have reserved trampolines for cases
+  // when a new callee address out of 128M range.
+  bool need_trampoline = (entry.rspec().type() == relocInfo::runtime_call_type) ?
```

Comment:
This isn't clear. Are you trying to prove that, even if the caller is moved as far as it possibly can be from the runtime target, the branch will still be in range? That is to say, all of the code in the current CodeHeap is within 128M of the target?

Reply:
> Are you trying to prove that, even if the caller is moved as far as it possibly can be from the runtime target, the branch will still be in range?

Yes, for calls into the non-nmethod segment.

> That is to say, all of the code in the current CodeHeap is within 128M of the target?

With the non-nmethod segment in the middle, it is 248M: 120M + 8M + 120M.